### PR TITLE
[AMDGPU] Introduce Code Object V6

### PR DESCRIPTION
**Pull request by Pierre-vh**
Introduce Code Object V6 in Clang, LLD, Flang and LLVM. This is the same as V5 except a new "generic version" flag can be present in EFLAGS. This is related to new generic targets that'll be added in a follow-up patch. It's also likely V6 will have new changes (possibly new metadata entries) added later.

Docs change are part of the follow-up patch #76955

### CODE REVIEW

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 1]
```diff
@@ -2585,7 +2585,7 @@ getAMDGPUCodeObjectArgument(const Driver &D, const llvm::opt::ArgList &Args) {
 void tools::checkAMDGPUCodeObjectVersion(const Driver &D,
                                          const llvm::opt::ArgList &Args) {
   const unsigned MinCodeObjVer = 4;
-  const unsigned MaxCodeObjVer = 5;
+  const unsigned MaxCodeObjVer = 6;
```
Comment (Pierre-vh) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:2588]: I'm wondering if we should print a warning when V6 is enabled (either here or in the backend) to note that it's in development and not ready yet? Something like "code object v6 is still experimental and not ready for production use"

File: lld/ELF/Arch/AMDGPU.cpp [hunk 2]
```diff
@@ -106,6 +107,25 @@ uint32_t AMDGPU::calcEFlagsV4() const {
...
+uint32_t AMDGPU::calcEFlagsV6() const {
+  uint32_t flags = calcEFlagsV4();
+
+  uint32_t genericVersion =
+      getEFlags(ctx.objectFiles[0]) & EF_AMDGPU_GENERIC_VERSION;
+
+  // Verify that all input files have compatible generic version.
+  for (InputFile *f : ArrayRef(ctx.objectFiles).slice(1)) {
+    if (genericVersion != (getEFlags(f) & EF_AMDGPU_GENERIC_VERSION)) {
+      // TODO: test
```
Comment (arsenm) [inline lld/ELF/Arch/AMDGPU.cpp:119]: This is tested?

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 3]
```diff
@@ -1557,140 +1559,98 @@ const EnumEntry<unsigned> ElfHeaderMipsFlags[] = {
   ENUM_ENT(EF_MIPS_ARCH_64R6, "mips64r6")
 };
 
+#define AMDGPU_MACH_ENUM_ENTS                                                  \
+  ENUM_ENT(EF_AMDGPU_MACH_NONE, "none"),                                       \
```
Comment (arsenm) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:1563]: First one differently indented from the rest

**Comment by Pierre-vh** [comment]
ping

**Comment by Pierre-vh** [comment]
ping

File: llvm/lib/ObjectYAML/ELFYAML.cpp [hunk 4]
```diff
@@ -620,6 +620,15 @@ void ScalarBitSetTraits<ELFYAML::ELF_EF>::bitset(IO &IO,
       BCase(EF_AMDGPU_FEATURE_XNACK_V3);
       BCase(EF_AMDGPU_FEATURE_SRAMECC_V3);
       break;
+    case ELF::ELFABIVERSION_AMDGPU_HSA_V6:
```
Comment (kzhuravl) [inline llvm/lib/ObjectYAML/ELFYAML.cpp:623]: I think we'd need to add a test for this?

File: llvm/include/llvm/Support/AMDGPUMetadata.h [hunk 5]
```diff
@@ -44,8 +44,15 @@ constexpr uint32_t VersionMajorV5 = 1;
 /// HSA metadata minor version for code object V5.
 constexpr uint32_t VersionMinorV5 = 2;
 
+/// HSA metadata major version for code object V6.
+constexpr uint32_t VersionMajorV6 = 1;
+/// HSA metadata minor version for code object V6.
+constexpr uint32_t VersionMinorV6 = 3;
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/include/llvm/Support/AMDGPUMetadata.h:50]: Not yet, but I assume we'll want to bundle some changes to the MD with V6 so it's better to update the version now, no?

File: llvm/lib/ObjectYAML/ELFYAML.cpp (hunk 4 above)
Comment (Pierre-vh) [reply llvm/lib/ObjectYAML/ELFYAML.cpp:623]: elf-headers.test already covers it

File: llvm/include/llvm/Support/AMDGPUMetadata.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (kzhuravl) [reply llvm/include/llvm/Support/AMDGPUMetadata.h:50]: As of now, there are no changes planned to the MD for v6

[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/include/llvm/Support/AMDGPUMetadata.h:50]: Do we just increment this number when there's a breaking metadata change? How does it work?

**Review by kzhuravl — APPROVED** [review]
Looks good, thanks!

**Pull request merged**